src_c/surface.c

```diff
@@ -2105,6 +2116,313 @@ surf_blits(pgSurfaceObject *self, PyObject *args, PyObject *keywds)
+                                    &dest_rect, src_rect, flags_numeric);
+
+            if (result != 0) {
+                errornum = BLITS_ERR_BLIT_FAIL;
+                goto on_error;
+            }
+            if (doreturn) {
+                PyList_SET_ITEM(ret, seq_counter, pgRect_New(&dest_rect));
+            }
+        }
+    }
+    else {
+        errornum = BLITS_ERR_SEQUENCE_REQUIRED;
+        goto on_error;
+    }
+
+    if (doreturn) {
+        return ret;
+    }
+    else {
+        Py_RETURN_NONE;
+    }
+
+on_error:
+    Py_XDECREF(retrect);
+    Py_XDECREF(iterator);
+    Py_XDECREF(ret);
+
+    switch (errornum) {
+        case BLITS_ERR_SEQUENCE_REQUIRED:
+            return RAISE(
+                PyExc_ValueError,
+                "blit_sequence should be iterator of (Surface, dest)");
+        case BLITS_ERR_DISPLAY_SURF_QUIT:
+            return RAISE(pgExc_SDLError, "display Surface quit");
+        case BLITS_ERR_SEQUENCE_SURF:
+            return RAISE(
+                PyExc_TypeError,
+                "First element of pairs (Surface, dest) in blit_sequence "
+                "must be a Surface.");
+        case BLITS_ERR_INVALID_DESTINATION:
+            return RAISE(PyExc_TypeError,
+                         "invalid destination position for blit");
+        case BLITS_ERR_INVALID_RECT_STYLE:
+            return RAISE(PyExc_TypeError, "Invalid rectstyle argument");
+        case BLITS_ERR_MUST_ASSIGN_NUMERIC:
+            return RAISE(PyExc_TypeError, "Must assign numeric values");
+        case BLITS_ERR_BLIT_FAIL:
+            return RAISE(
+                PyExc_TypeError,
+                "Blit failed (probably the flag used does not exist)");
+        case BLITS_ERR_PY_EXCEPTION_RAISED:
+            return NULL; /* Raising a previously set exception */
+        case BLITS_ERR_SOURCE_NOT_SURFACE:
+            return RAISE(PyExc_TypeError, "Source objects must be a Surface");
+        case UBLITS_ERR_TUPLE_REQUIRED:
+            return RAISE(
+                PyExc_ValueError,
+                "Blit_sequence item should be a tuple of (Surface, dest)");
+        case UBLITS_ERR_INSUFFICIENT_ARGS:
+            return RAISE(PyExc_ValueError,
+                         "Function requires positional arguments in the "
+                         "order: blit_sequence, special_flags, doreturn");
+        case UBLITS_ERR_FLAG_NOT_NUMERIC:
+            return RAISE(PyExc_ValueError,
+                         "The special_flags parameter must be an int");
+        case UBLITS_ERR_DORETURN_NOT_NUMERIC:
+            return RAISE(
+                PyExc_ValueError,
+                "The doreturn parameter must either be a bool or numeric");
+    }
+    return RAISE(PyExc_TypeError, "Unknown error");
+}
+#else
+
+static PyObject *
+surf_ublits(pgSurfaceObject *self, PyObject *_null)
+{
+    return RAISE(PyExc_NotImplementedError,
+                 "ublits is not compatible with Python 3.6 or older");
+}
+
```

Comment:
I'm not a fan of this, I believe it's best if all functions are consistently defined on all versions we support.

I propose that *only* on python 3.6, we have a wrapper `METH_VARARGS` function that calls the `METH_FASTCALL` function (which should be defined regardless of python version, with full implementation), and this wrapper should be like

Here is some (untested) code that demonstrates what I'm thinking

```c
#if PY_VERSION_HEX < 0x03070000
static PyObject *
surf_ublits_wrapper(PyObject *self, PyObject *args)
{
    Py_ssize_t nargs = PyTuple_GET_SIZE(args);
    PyObject *const *vector_args = PyMem_New(PyObject *const *, nargs);
    if (!vector_args) {
        return PyErr_NoMemory();
    }

    PyObject *ret = surf_ublits(self, vector_args, nargs);
    PyMem_Free(vector_args);
    return ret;
}
#endif
```

Reply:
I agree with you in theory, but in practice, how many people even use `blits`?

And how many of those people actually update pygame and read the release notes? And how many of those people haven't updated their python installation since 2016/2017? I don't see this affecting [m]any people.

I think it would be fine to do this for a specialized function on an unsupported version of python that we should drop support for soon.

Reply:
I see where you are coming from, but in this case the little patch I suggested above is simple enough (atleast I think so) so for the sake of consistency and cleanliness IMHO there is no harm in using it. Obviously this codepath would miss out on the performance enhancements of 3.7 `METH_FASTCALL`, but I think that's fine for the same reasons you give (very little number of users affected by this)

Better to have a slower specialised function with just some more extra lines of code, than one that won't even work and give that one user a hard time

Reply:
So what i stick with the current version?

Reply:
> I'm not a fan of this, I believe it's best if all functions are consistently defined on all versions we support.
> 
> I propose that _only_ on python 3.6, we have a wrapper `METH_VARARGS` function that calls the `METH_FASTCALL` function (which should be defined regardless of python version, with full implementation), and this wrapper should be like
> 
> Here is some (untested) code that demonstrates what I'm thinking
> 
> ```c
> #if PY_VERSION_HEX < 0x03070000
> static PyObject *
> surf_ublits_wrapper(PyObject *self, PyObject *args)
> {
>     Py_ssize_t nargs = PyTuple_GET_SIZE(args);
>     PyObject *const *vector_args = PyMem_New(PyObject *const *, nargs);
>     if (!vector_args) {
>         return PyErr_NoMemory();
>     }
> 
>     PyObject *ret = surf_ublits(self, vector_args, nargs);
>     PyMem_Free(vector_args);
>     return ret;
> }
> #endif
> ```

this version you propose won't even be possible anymore, as the function now always returns None.

Reply:
I don't see how it's not possible? Afterall python `None` is still `Py_None` (which is a `PyObject *`) at the C level

Reply:
> as the function now always returns None.

Did you forget to push a commit or something? The function still has optional returns on this end.

Reply:
It's not finished 100% i didn't push it because i'm having an exam today.

Reply:
i mean it should work in theory, i'll push it and see how it goes

Reply:
i did something strange with i hope it worked

Reply:
Can we consider this resolved?

Reply:
You didn't resolve the issue though?